cmd/soroban-rpc/internal/feewindow/feewindow.go

```diff
@@ -0,0 +1,178 @@
+	fw.lock.Lock()
+	defer fw.lock.Unlock()
+	_, err := fw.feesPerLedger.Append(fees)
+	if err != nil {
+		return err
+	}
+
+	var allFees []uint64
+	for i := uint32(0); i < fw.feesPerLedger.Len(); i++ {
+		allFees = append(allFees, fw.feesPerLedger.Get(i).BucketContent...)
+	}
+	fw.distribution = computeFeeDistribution(allFees)
+
+	return nil
+}
+
+func computeFeeDistribution(fees []uint64) FeeDistribution {
+	if len(fees) == 0 {
+		return FeeDistribution{}
+	}
+	slices.Sort(fees)
+	mode := fees[0]
+	lastVal := fees[0]
+	maxRepetitions := 0
+	localRepetitions := 0
+	for i := 1; i < len(fees); i++ {
+		if fees[i] == lastVal {
+			localRepetitions += 1
+			continue
+		}
+
+		// new cluster of values
+
+		if localRepetitions > maxRepetitions {
+			maxRepetitions = localRepetitions
+			mode = fees[i]
+		}
+		lastVal = fees[i]
+		localRepetitions = 0
+	}
+	count := uint64(len(fees))
+	// nearest-rank percentile
+	percentile := func(p uint64) uint64 {
+		// ceiling(p*count/100)
+		kth := ((p * count) + 100 - 1) / 100
+		return fees[kth-1]
+	}
+	return FeeDistribution{
+		Max:   fees[len(fees)-1],
+		Min:   fees[0],
+		Mode:  mode,
+		P10:   percentile(10),
+		P20:   percentile(20),
+		P30:   percentile(30),
+		P40:   percentile(40),
+		P50:   percentile(50),
+		P60:   percentile(60),
+		P70:   percentile(70),
+		P80:   percentile(80),
+		P90:   percentile(90),
+		P95:   percentile(95),
+		P99:   percentile(99),
+		Count: count,
+	}
+}
+
+func (fw *FeeWindow) GetFeeDistribution() FeeDistribution {
+	fw.lock.RLock()
+	defer fw.lock.RUnlock()
+	return fw.distribution
+}
+
+type FeeWindows struct {
+	SorobanInclusionFeeWindow *FeeWindow
+	ClassicFeeWindow          *FeeWindow
+	networkPassPhrase         string
+}
+
+func NewFeeWindows(classicRetention uint32, sorobanRetetion uint32, networkPassPhrase string) *FeeWindows {
+	return &FeeWindows{
+		SorobanInclusionFeeWindow: NewFeeWindow(sorobanRetetion),
+		ClassicFeeWindow:          NewFeeWindow(classicRetention),
+		networkPassPhrase:         networkPassPhrase,
+	}
+}
+
+func (fw *FeeWindows) IngestFees(meta xdr.LedgerCloseMeta) error {
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(fw.networkPassPhrase, meta)
+	if err != nil {
+		return err
+	}
+	var sorobanInclusionFees []uint64
+	var classicFees []uint64
+	for {
+		tx, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		ops := tx.Envelope.Operations()
+		if len(ops) == 1 {
+			switch ops[0].Body.Type {
+			case xdr.OperationTypeInvokeHostFunction, xdr.OperationTypeExtendFootprintTtl, xdr.OperationTypeRestoreFootprint:
+				if tx.Envelope.V1 == nil || tx.Envelope.V1.Tx.Ext.SorobanData != nil {
+					// this shouldn't happen
+					continue
+				}
+				inclusionFee := uint64(tx.Envelope.V1.Tx.Fee) - uint64(tx.Envelope.V1.Tx.Ext.SorobanData.ResourceFee)
+				classicFees = append(sorobanInclusionFees, inclusionFee)
```

Comment:
```
SorobanInclusionFee FeeDistribution `json:"sorobanInclusionFee"` => Soroban Inclusion Fee Distribution
InclusionFee        FeeDistribution `json:"inclusionFee"` => Classic Inclusion Fee Distribution
```
These two inclusion fees are separate and are unrelated distributions.

Reply:
currently `classicFees` only contains fees for soroban transactions because the if statement and the switch statement filters away transactions which have more than 1 operation and transactions which consist of non soroban operations

Reply:
@tamirms that's a bug, it should be:

```suggestion
				sorobanInclusionFees = append(sorobanInclusionFees, inclusionFee)
```

Fixed